src/guardian/staking-module-guard/staking-module-guard.service.ts

```diff
@@ -300,6 +313,98 @@ export class StakingModuleGuardService {
     await this.guardianMessageService.sendPauseMessage(pauseMessage);
   }
 
+  public async isVettedUnusedKeysValid(
+    stakingModuleData: StakingModuleData,
+    blockData: BlockData,
+  ): Promise<boolean> {
+    const { blockNumber, depositRoot } = blockData;
+    const { nonce, stakingModuleId, lastChangedBlockHash } = stakingModuleData;
+    const lastContractsState =
+      this.lastContractsStateByModuleId[stakingModuleId];
+
+    if (
+      lastContractsState &&
+      lastChangedBlockHash === lastContractsState.lastChangedBlockHash &&
+      lastContractsState.invalidKeysFound
+    ) {
+      // if found invalid keys on previous iteration and lastChangedBlockHash returned by kapi was not changed
+      // we dont need to validate again, but we still need to skip deposits until problem will not be solved
+      this.logger.error(
+        'LastChangedBlockHash was not changed and on previous iteration we found invalid keys, skip until solving problem ',
+      );
+
+      this.lastContractsStateByModuleId[stakingModuleId] = {
+        nonce,
+        depositRoot,
+        blockNumber,
+        lastChangedBlockHash,
+        invalidKeysFound: true,
+      };
+
+      return false;
+    }
+
+    if (
+      !lastContractsState ||
+      lastChangedBlockHash !== lastContractsState.lastChangedBlockHash
+    ) {
+      // keys was changed or it is a first attempt, need to validate again
+      const invalidKeys = await this.getInvalidKeys(
+        stakingModuleData,
+        blockData,
+      );
+
+      // if found invalid keys, update state and exit
+      if (invalidKeys.length) {
+        this.logger.error(
+          'Found invalid keys, will skip deposits until solving problem',
+        );
+        this.guardianMetricsService.incrInvalidKeysEventCounter();
+        // save info about invalid keys in cache
+        this.lastContractsStateByModuleId[stakingModuleId] = {
+          nonce,
+          depositRoot,
+          blockNumber,
+          lastChangedBlockHash,
+          invalidKeysFound: true,
+        };
+
+        return false;
+      }
+
+      // keys are valid, state will be updated later
+      return true;
+    }
+
+    return true;
+  }
+
+  public async getInvalidKeys(
+    stakingModuleData: StakingModuleData,
+    blockData: BlockData,
+  ): Promise<{ key: string; depositSignature: string }[]> {
+    this.logger.log('Start keys validation', {
+      keysCount: stakingModuleData.vettedUnusedKeys.length,
+      moduleId: stakingModuleData.stakingModuleId,
+    });
+    const validationTimeStart = performance.now();
+    const invalidKeysList = await this.keysValidationService.findInvalidKeys(
```

Comment:
it is not very convenient to read the findInvalidKeys and getInvalidKeys methods. While reading it seems that this is a request to the server, but not validation. Perhaps they should be renamed?

Reply:
Good names usually don't come easily. For 'find', it could be 'validate', but there's also a cache, so maybe 'find' is more suitable. As for 'getInvalidKeys', I'm not sure how to rename it right now.